p2p/peer.go

```diff
@@ -189,35 +257,53 @@ func (p *peer) String() string {
 // SetLogger implements BaseService.
 func (p *peer) SetLogger(l log.Logger) {
 	p.Logger = l
-	p.mconn.SetLogger(l)
 }
 
 // OnStart implements BaseService.
 func (p *peer) OnStart() error {
-	if err := p.BaseService.OnStart(); err != nil {
-		return err
+	// Open streams for all reactors.
+	p.streams = make(map[byte]transport.Stream)
+	for streamID, info := range p.streamInfoByStreamID {
+		var d transport.StreamDescriptor
+		descs := info.reactor.StreamDescriptors()
+		for _, desc := range descs {
+			if desc.StreamID() == streamID {
+				d = desc
+				break
+			}
+		}
+		stream, err := p.peerConn.OpenStream(streamID, d)
+		if err != nil {
+			return fmt.Errorf("opening stream %v: %w", streamID, err)
+		}
+		p.streams[streamID] = stream
 	}
 
-	if err := p.mconn.Start(); err != nil {
-		return err
+	// Start the MConnection if it's an MConnection.
+	// NOTE: we do not start the MConnection until all the streams are registered.
+	if mconn, ok := p.peerConn.Conn.(*tcpconn.MConnection); ok {
+		if err := mconn.Start(); err != nil {
+			return fmt.Errorf("starting MConnection: %w", err)
+		}
 	}
 
-	go p.metricsReporter()
+	go p.eventLoop()
+
 	return nil
 }
 
 // FlushStop mimics OnStop but additionally ensures that all successful
-// .Send() calls will get flushed before closing the connection.
-//
-// NOTE: it is not safe to call this method more than once.
+// Send calls will get flushed before closing the connection.
 func (p *peer) FlushStop() {
-	p.mconn.FlushStop() // stop everything and close the conn
+	if err := p.FlushAndClose("stopping peer"); err != nil {
+		p.Logger.Error("Close", "err", err)
+	}
 }
 
 // OnStop implements BaseService.
 func (p *peer) OnStop() {
-	if err := p.mconn.Stop(); err != nil { // stop everything and close the conn
-		p.Logger.Debug("Error while stopping peer", "err", err)
+	if err := p.Close("stopping peer"); err != nil {
+		p.Logger.Error("Close", "err", err)
```

Comment:
We are increasing log level here. Afraid it might create too much noise.

Reply:
But it's only executed if closing underlying `MConnection` fails